rust/src/lib/nm/query_apply/route.rs

```diff
@@ -1,8 +1,70 @@
 // SPDX-License-Identifier: Apache-2.0
 
+use crate::{RouteEntry, RouteType};
+
 use super::super::nm_dbus::{NmConnection, NmIpRoute};
 
 const DEFAULT_TABLE_ID: u32 = 254; // main route table ID
+const NM_EMPTY_NEXT_HOP_ADDRV4: &str = "0.0.0.0";
+const NM_EMPTY_NEXT_HOP_ADDRV6: &str = "::";
+const NM_DEFAULT_IPV4_METRIC: u32 = 0;
+
+// Both `nm_dbus` are supposed to be a standalone crates,
+// `RouteEntry` and `NmIpRoute` are defined outside of `nm` crate,
+// Hence we cannot do `impl From<NmIpRoute> for RouteEntry` here.
+
+fn nm_route_to_route_entry(nm_route: &NmIpRoute) -> RouteEntry {
+    let mut route = RouteEntry {
+        state: None,
+        next_hop_iface: None,
+        destination: nm_route.dest.clone(),
+        next_hop_addr: {
+            if let Some(next_hop) = nm_route.next_hop.as_deref() {
+                if next_hop != NM_EMPTY_NEXT_HOP_ADDRV4
+                    && next_hop != NM_EMPTY_NEXT_HOP_ADDRV6
+                {
+                    Some(next_hop.to_string())
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        },
+        metric: nm_route.metric.as_ref().and_then(|metric| {
+            if *metric == NM_DEFAULT_IPV4_METRIC {
+                None
+            } else {
+                Some(*metric as i64)
+            }
+        }),
+        table_id: nm_route.table.or(Some(DEFAULT_TABLE_ID)),
```

Comment:
This is not necessarely true. If the existing NM connection has ipv4/6.route-table=100, then the route table won't default to 254.